### PR TITLE
Service param added to agency controller to limit results by service_category

### PR DESCRIPTION
**Pull request by PhilNorman2**
Added "service" query param to the agency controller index method.  This parameter maps to the service_category (primary key of service_categories).  For example limiting an agencies request with a service param of 20 returns only agencies that have events with a service_category_name that corresponds to that service param.  In this case, the service param returns only the agencies that have events with a service_category_name of "Produce".

Note:  This action actually involves two queries.  Agency.by_service_category limits the results to the Agencies that contain at least one event with a matching service_category.  Event.by_service_category limits the results within each Agency to those with a matching service_category.  Event.by_service_category is set up in the Agency serializer.  
```
   has_many :events

   def events
    # required to limit events by agencies controller service param
    # because agency version of query limits by agencies only
    if (service = instance_options[:service])
      return object.events.by_service_category(service)
    end

    object.events
  end
```
The Event version alone will provide the correct results but the performance when just using the service query param is terrible.  Using both queries only adds a few milliseconds when used along with the zip_code param.  Based on these performance results I included both queries in the action.

Example request

http://localhost:8888/api/agencies?service=20

Example response (truncated)

<pre>
{
    "agencies": [
        {
            "id": 12,
            "address": "15500 S WATERLOO RD",
            "city": "CLEVELAND",
            "state": "OH",
            "zip": "44110",
            "phone": "216.738.2265",
            "name": "Greater Cleveland Food Bank",
            "nickname": "Greater Cleveland Food Bank",
            "estimated_distance": "",
            "events": [
                {
                    "id": 42,
                    "address": "15500 S WATERLOO RD ",
                    "city": "CLEVELAND",
                    "state": "OH",
                    "zip": "44110",
                    "agency_id": 12,
                    "latitude": "41.5669006",
                    "longitude": "-81.5724759",
                    "name": "MOBILE PAN",
                    "service": "Produce",
                    "estimated_distance": "",
                    "exception_note": "",
                    "event_details": "",
                    "agency_name": "Greater Cleveland Food Bank",
                    "event_dates": [],
                    "forms": [
                        {
                            "id": 91,
                            "display_age_adult": 18,
                            "display_age_senior": 60
                        }
                    ]
                }
            ]
        },
        {
            "id": 1362,
            "address": "301 N Main",
            "city": "BETHESDA",
            "state": "OH",
            "zip": "43719",
            "phone": "740-298-0076",
            "name": "Belmont Mobile Market",
            "nickname": "Union Local Schools",
            "estimated_distance": "",
            "events": [
                {
                    "id": 1993,
                    "address": "301 N Main  ",
                    "city": "BETHESDA",
                    "state": "OH",
                    "zip": "43719",
                    "agency_id": 1362,
                    "latitude": "40.0185289",
                    "longitude": "-81.07400753",
                    "name": "Mobile Market",
                    "service": "Produce",
                    "estimated_distance": "",
                    "exception_note": "",
                    "event_details": "",
                    "agency_name": "Belmont Mobile Market",
                    "event_dates": [
                        {
                            "id": 7418,
                            "event_id": 1993,
                            "capacity": 10,
                            "accept_walkin": 1,
                            "accept_reservations": 0,
                            "accept_interest": 0,
                            "start_time": "10 AM",
                            "end_time": "12 PM",
                            "date": "2020-08-06"
                        },
                        {
                            "id": 7419,
                            "event_id": 1993,
                            "capacity": 10,
                            "accept_walkin": 1,
                            "accept_reservations": 0,
                            "accept_interest": 0,
                            "start_time": "10 AM",
                            "end_time": "12 PM",
                            "date": "2020-08-20"
                        }
                    ],
                    "forms": [
                        {
                            "id": 91,
                            "display_age_adult": 18,
                            "display_age_senior": 60
                        }
                    ]
                }
            ]
        }
    ...
</pre>

### CODE REVIEW

**Comment by nvengal** [comment]
This isn't really a blocking review, but more a point of discussion since it seems like we're relying more and more on the pattern of passing instance options into the serializer and performing queries within the serialization process. I don't think that's a very good pattern, for the sake of testability/readability and the performance concerns that you've been noticing (though there are things we can do to mitigate the performance concerns).

Here's my concern. We're shoving more and more things into this single endpoint. It's like we're moving towards encapsulating the whole api behind just the `/agencies` route as we add more options and serialize more objects into the payload.

Specifically, for this pull request, does the agencies api really need to filter events? (https://github.com/midohiofoodbank/freshtrak-pantry-finder-api/blob/b9909a37a0e61573730f83a86c42d950bac0dc35/app/serializers/agency_serializer.rb#L12-L20)

It seems to me like event filtering only needs to happen on the `/events` endpoints. When you get the events for an agency through the agencies, it's reasonable (and I would be least surprised) for all the events for that agency to be included. If we chose to hide some of those in our frontend application that's fine, but do they needed to be hidden at the api level?

Put differently, the way the filtering works in this pr, expects the consumer of this api to have in-depth knowledge of our data structure. They should know that service types are connected to an agency through events and that's why only some of the events show up when you filter by service type. That's a strange thing to expect a user to know, especially when we're considering opening this api up to external consumers.

In summary, I think it's more than reasonable to filter agencies by service type, but that doesn't mean we need to filter nested events in the same way. If you want the api to filter your events, use the `/api/events` endpoint.

**Comment by PhilNorman2** [comment]
Very good points.   I believe in terms of intent, our default case is to
filter by events because that is what our users care about.  For example,
 I don't doubt the event_date param gives the results David was expecting
through the Agencies resource.  I am concerned that our zip_code logic may
be doing the same thing, but would need to investigate.  Conceptually, I
think we could move all these scopes to the Event model and the client
could use the events resource for their request.  It already provides the
Agency name.  Just an idea for discussion.

**Comment by PhilNorman2** [comment]
FYI: I performed multiple console Active record queries on multiple zip codes against the Agency and Event and compared with JSON output this morning and did not find cases of events in an agency outside of the event service geography (didn't match the zip_code param).  I am guessing we don't have Agencies with events outside of their event service geography, or if we do it is rare.

**Comment by PhilNorman2** [comment]
The team met to step back and look at the filtering capability for service_category and others in the future.  The following options were discussed to filter results:

(1) from the events endpoint
(2) in the UI
(3) hybrid approach, first from the agencies API to narrow results and then from the UI.

(1) was ruled out because we may need additional Agencies data.  Also even if we used the events endpoint we would have to search by the event_date endpoint to filter by event_date which would be complicated
(2) was chosen over (3) primarily for simplicity sake.  

Basically we are establishing a pattern of separating search from filtering where filtering is done in UI.

As a result of this decision PR #44 is being closed.  Another PR will be created to add a service_category serializer to render the id and service_category_name.

**Comment by davidpickering** [comment]
@PhilNorman2 a closed PR can be a sad day, thank you for leading good thinking and helping keep us moving towards a pure-er implementation. Every line adds up and
